molgenis-data-index/src/main/java/org/molgenis/data/index/IndexActionRegisterServiceImpl.java

```diff
@@ -109,85 +110,127 @@ public synchronized void register(EntityType entityType, String entityId)
+		dataService.add(INDEX_ACTION, indexActions.stream());
 	}
 
 	/**
-	 * Filter all unnecessary index actions
+	 * Determines which IndexActions are necessary to bring the index up to date with the current transaction.
 	 *
-	 * @return Set<IndexAction>
+	 * @return List<IndexAction> List of IndexActions that are necessary
 	 */
-	private Set<IndexAction> filterUnnecessaryIndexActions()
+	private List<IndexAction> determineNecessaryActions()
 	{
-		// 1. add all referencing entities
-		Set<IndexAction> allIndexAction = getIndexActionsForCurrentTransaction().stream()
-				.flatMap(this::addReferencingEntities).collect(toSet());
-
-		// 2. Filter excluded entities
-		Set<IndexAction> indexActionWithoutExcluded = allIndexAction.stream()
-				.filter(indexAction -> !excludedEntities.contains(indexAction.getEntityTypeId())).collect(toSet());
-
-		// 3. Find all entities names of actions where no row is specified
-		Set<String> entityFullIds = indexActionWithoutExcluded.stream()
-				.filter(indexAction -> indexAction.getEntityId() == null).map(IndexAction::getEntityTypeId)
-				.collect(toSet());
-
-		// 4. Filter all row index actions from list
-		return indexActionWithoutExcluded.stream()
-				.filter(indexAction -> (indexAction.getEntityId() == null) || !entityFullIds
-						.contains(indexAction.getEntityTypeId())).collect(toSet());
+		ImmutableSet<IndexAction> indexActions = copyOf(getIndexActionsForCurrentTransaction());
+		if (indexActions.isEmpty())
+		{
+			return Collections.emptyList();
+		}
+		Stopwatch sw = Stopwatch.createStarted();
+		List<IndexAction> result = determineNecessaryActionsInternal(indexActions,
+				indexActions.iterator().next().getIndexActionGroup(), new IndexDependencyModel(getEntityTypes()));
+		LOG.debug("Determined necessary actions in {}", sw);
+		return result;
 	}
 
 	/**
-	 * Add for all referencing entities an index action
+	 * Retrieves all {@link EntityType}s.
+	 * Queryies in pages of size ENTITY_FETCH_PAGE_SIZE so that results can be cached.
+	 * Uses a {@link Fetch} that specifies all fields needed to determine the necessary index actions.
 	 *
-	 * @return Stream<IndexAction>
+	 * @return List containing all {@link EntityType}s.
 	 */
-	private Stream<IndexAction> addReferencingEntities(IndexAction indexAction)
+	private List<EntityType> getEntityTypes()
 	{
-		EntityType entityType = dataService.getEntityType(indexAction.getEntityTypeId());
-		if (entityType == null) // When entity is deleted the entityType cannot be retrieved
+		Fetch fetch = new Fetch();
+		fetch.field(ID);
+		fetch.field(IS_ABSTRACT);
+		fetch.field(INDEXING_DEPTH);
+
+		Fetch extendsFetch = new Fetch();
+		extendsFetch.field(ID);
+		fetch.field(EXTENDS, extendsFetch);
+
+		Fetch attributesFetch = new Fetch();
+		Fetch refEntityFetch = new Fetch();
+		refEntityFetch.field(ID);
+		attributesFetch.field(AttributeMetadata.REF_ENTITY_TYPE, refEntityFetch);
+		fetch.field(ATTRIBUTES, attributesFetch);
```

Comment:
We could put the creation of the Fetch in a static initializer block so we don't need to create it each time. This will also clean up the method a bit